File: src/subtypes.ts
// Reverse subtype index: super type symbol -> the type declarations that
// directly extend/implement it, built in one workspace pass and memoized on the
// program generation. Backs textDocument/implementation, transitive
// implementation counts in code lenses, and (later) type/call hierarchy.

import { forEachChild } from "./parser.ts";
import type { Program } from "./program.ts";
import { resolveTypeEntityName } from "./resolver.ts";
import { isSyntheticUri } from "./workspace.ts";
import {
  type Identifier,
  type MethodDeclaration,
  type Node,
  type Symbol,
  SyntaxKind,
  type TypeNode,
  type TypeReference,
} from "./types.ts";

const TYPE_DECLARATIONS = new Set<SyntaxKind>([
  SyntaxKind.ClassDeclaration,
  SyntaxKind.InterfaceDeclaration,
  SyntaxKind.EnumDeclaration,
  SyntaxKind.RecordDeclaration,
]);

export interface SubtypeIndex {
  /** Direct subtypes (declarations whose extends/implements names the symbol). */
  directSubtypesOf(superType: Symbol): readonly Symbol[];
  /** All transitive subtypes, the direct ones first (BFS order, no duplicates). */
  allSubtypesOf(superType: Symbol): Symbol[];
}

function superTypeNodes(node: Node): readonly TypeNode[] {
  const d = node as {
    extendsType?: TypeNode;
    extendsTypes?: readonly TypeNode[];
    implementsTypes?: readonly TypeNode[];
  };
  return [
    ...(d.extendsType ? [d.extendsType] : []),
    ...(d.extendsTypes ?? []),
    ...(d.implementsTypes ?? []),
  ];
}

const indexCache = new WeakMap<Program, { generation: number; index: SubtypeIndex }>();

export function getSubtypeIndex(program: Program): SubtypeIndex {
  const generation = program.getGeneration();
  const cached = indexCache.get(program);
  if (cached && cached.generation === generation) return cached.index;

  const direct = new Map<Symbol, Symbol[]>();
  for (const uri of program.getAllUris()) {
    if (isSyntheticUri(uri)) continue; // stub types never extend user code
    const sourceFile = program.getSourceFile(uri);
    if (!sourceFile) continue;
    const visit = (node: Node): void => {
      if (TYPE_DECLARATIONS.has(node.kind) && node.symbol) {
        for (const superType of superTypeNodes(node)) {
          if (superType.kind !== SyntaxKind.TypeReference) continue;
          const superSymbol = resolveTypeEntityName(
            (superType as TypeReference).typeName,
            node,
            program,
          );
          if (!superSymbol) continue;
          const list = direct.get(superSymbol);
          if (list) list.push(node.symbol);
          else direct.set(superSymbol, [node.symbol]);
        }
      }
      forEachChild(node, child => {
        visit(child);
        return undefined;
      });
    };
    visit(sourceFile);
  }

  const index: SubtypeIndex = {
    directSubtypesOf: superType => direct.get(superType) ?? [],
    allSubtypesOf(superType) {
      const seen = new Set<Symbol>();
      const queue = [...(direct.get(superType) ?? [])];
      const out: Symbol[] = [];
      while (queue.length > 0) {
        const next = queue.shift()!;
        if (seen.has(next)) continue;
        seen.add(next);
        out.push(next);
        queue.push(...(direct.get(next) ?? []));
      }
      return out;
    },
  };
  indexCache.set(program, { generation, index });
  return index;
}

/**
 * The concrete method declarations implementing/overriding `method` (matched by
 * name and arity) in the transitive subtypes of its declaring type.
 */
export function findMethodImplementations(
  method: MethodDeclaration,
  program: Program,
): MethodDeclaration[] {
  const owner = method.parent?.symbol;
  if (!owner) return [];
  const result: MethodDeclaration[] = [];
  for (const subtype of getSubtypeIndex(program).allSubtypesOf(owner)) {
    const declaration = subtype.valueDeclaration ?? subtype.declarations?.[0];
    const members = (declaration as { members?: readonly Node[] } | undefined)?.members ?? [];
    for (const member of members) {
      if (member.kind !== SyntaxKind.MethodDeclaration) continue;
      const candidate = member as MethodDeclaration;
      if (
        candidate.body &&
        candidate.name.text === method.name.text &&
        candidate.parameters.length === method.parameters.length
      ) {
        result.push(candidate);
      }
    }
  }
  return result;
}

/** The name node of a symbol's primary declaration, for jump targets. */
export function declarationName(symbol: Symbol): Identifier | undefined {
  const declaration = symbol.valueDeclaration ?? symbol.declarations?.[0];
  return (declaration as { name?: Identifier } | undefined)?.name;
}

File: togo/internal/services/subtypes.go
package services

// Reverse subtype index: super type symbol -> the type declarations that
// directly extend/implement it, built in one workspace pass and memoized on the
// program generation. Backs textDocument/implementation, transitive
// implementation counts in code lenses, and (later) type/call hierarchy.
// Port of src/services/subtypes.ts.

import "github.com/nikeee/cappu/internal/compiler"

// SubtypeIndex is a reverse map from super types to their declared subtypes.
type SubtypeIndex struct {
	direct map[*compiler.Symbol][]*compiler.Symbol
}

// DirectSubtypesOf returns declarations whose extends/implements names the symbol.
func (s *SubtypeIndex) DirectSubtypesOf(superType *compiler.Symbol) []*compiler.Symbol {
	return s.direct[superType]
}

// AllSubtypesOf returns all transitive subtypes, direct ones first (BFS order).
func (s *SubtypeIndex) AllSubtypesOf(superType *compiler.Symbol) []*compiler.Symbol {
	seen := map[*compiler.Symbol]bool{}
	queue := append([]*compiler.Symbol{}, s.direct[superType]...)
	var out []*compiler.Symbol
	for len(queue) > 0 {
		next := queue[0]
		queue = queue[1:]
		if seen[next] {
			continue
		}
		seen[next] = true
		out = append(out, next)
		queue = append(queue, s.direct[next]...)
	}
	return out
}

type subtypeCacheEntry struct {
	generation compiler.Generation
	index      *SubtypeIndex
}

var subtypeCache = map[*compiler.Program]subtypeCacheEntry{}

// GetSubtypeIndex builds (or returns the memoized) subtype index for a program.
func GetSubtypeIndex(program *compiler.Program) *SubtypeIndex {
	generation := program.GetGeneration()
	if cached, ok := subtypeCache[program]; ok && cached.generation == generation {
		return cached.index
	}

	direct := map[*compiler.Symbol][]*compiler.Symbol{}
	for _, uri := range program.GetAllUris() {
		if compiler.IsSyntheticURI(uri) {
			continue // stub types never extend user code
		}
		sourceFile := program.GetSourceFile(uri)
		if sourceFile == nil {
			continue
		}
		var visit compiler.Visitor
		visit = func(node *compiler.Node) bool {
			if isTypeDeclaration(node) && node.Symbol != nil {
				for _, superType := range superTypeNodes(node) {
					if superType.Kind != compiler.TypeReference {
						continue
					}
					superSymbol := compiler.ResolveTypeEntityName(superType.AsTypeReference().TypeName, node, program)
					if superSymbol == nil {
						continue
					}
					direct[superSymbol] = append(direct[superSymbol], node.Symbol)
				}
			}
			node.ForEachChild(visit)
			return false
		}
		visit(sourceFile)
	}

	index := &SubtypeIndex{direct: direct}
	subtypeCache[program] = subtypeCacheEntry{generation: generation, index: index}
	return index
}

// FindMethodImplementations returns the concrete method declarations
// implementing/overriding method (matched by name and arity) in the transitive
// subtypes of its declaring type.
func FindMethodImplementations(method *compiler.Node, program *compiler.Program) []*compiler.Node {
	if method.Parent == nil || method.Parent.Symbol == nil {
		return nil
	}
	owner := method.Parent.Symbol
	md := method.AsMethodDeclaration()
	var result []*compiler.Node
	for _, subtype := range GetSubtypeIndex(program).AllSubtypesOf(owner) {
		declaration := symbolDeclaration(subtype)
		if declaration == nil {
			continue
		}
		for _, member := range declMembers(declaration) {
			if member.Kind != compiler.MethodDeclaration {
				continue
			}
			cm := member.AsMethodDeclaration()
			if cm.Body != nil && cm.Name.AsIdentifier().Text == md.Name.AsIdentifier().Text &&
				cm.Parameters.Len() == md.Parameters.Len() {
				result = append(result, member)
			}
		}
	}
	return result
}

// DeclarationName returns the name node of a symbol's primary declaration.
func DeclarationName(symbol *compiler.Symbol) *compiler.Node {
	declaration := symbolDeclaration(symbol)
	if declaration == nil {
		return nil
	}
	return declName(declaration)
}
